litmus-portal/graphql-server/pkg/chaoshub/service.go

```diff
@@ -15,48 +15,63 @@ import (
 	"github.com/litmuschaos/litmus/litmus-portal/graphql-server/pkg/chaoshub/handler"
 	chaosHubOps "github.com/litmuschaos/litmus/litmus-portal/graphql-server/pkg/chaoshub/ops"
-	dbOperationsChaosHub "github.com/litmuschaos/litmus/litmus-portal/graphql-server/pkg/database/mongodb/chaoshub"
 	dbSchemaChaosHub "github.com/litmuschaos/litmus/litmus-portal/graphql-server/pkg/database/mongodb/chaoshub"
+	"go.mongodb.org/mongo-driver/bson"
 )
 
 const (
 	timeInterval = 6 * time.Hour
 	defaultPath  = "/tmp/version/"
 )
 
-// AddChaosHub is used for Adding a new ChaosHub
-func AddChaosHub(ctx context.Context, chaosHub model.CreateChaosHubRequest) (*model.ChaosHub, error) {
+type Service interface {
+	AddChaosHub(ctx context.Context, chaosHub model.CreateChaosHubRequest) (*model.ChaosHub, error)
+	AddRemoteChaosHub(ctx context.Context, chaosHub model.CreateRemoteChaosHub) (*model.ChaosHub, error)
+	SaveChaosHub(ctx context.Context, chaosHub model.CreateChaosHubRequest) (*model.ChaosHub, error)
+	SyncHub(ctx context.Context, hubID string, projectID string) (string, error)
+	UpdateChaosHub(ctx context.Context, chaosHub model.UpdateChaosHubRequest) (*model.ChaosHub, error)
+	DeleteChaosHub(ctx context.Context, hubID string, projectID string) (bool, error)
+	ListCharts(ctx context.Context, hubName string, projectID string) ([]*model.Chart, error)
+	GetHubExperiment(ctx context.Context, request model.ExperimentRequest) (*model.Chart, error)
+	ListHubStatus(ctx context.Context, projectID string) ([]*model.ChaosHubStatus, error)
+	GetYAMLData(request model.ExperimentRequest) (string, error)
+	GetExperimentManifestDetails(ctx context.Context, request model.ExperimentRequest) (*model.ExperimentDetails, error)
+	ListPredefinedWorkflows(hubName string, projectID string) ([]*model.PredefinedWorkflowList, error)
+	GetPredefinedExperimentYAMLData(request model.ExperimentRequest) (string, error)
+	IsChaosHubAvailable(ctx context.Context, hubName string, projectID string) (bool, error)
+	GetAllHubs(ctx context.Context) ([]*model.ChaosHub, error)
+	RecurringHubSync()
+}
 
-	IsExist, err := IsChaosHubAvailable(ctx, chaosHub.HubName, chaosHub.ProjectID)
-	if err != nil {
-		return nil, err
-	}
-	if IsExist == true {
-		return nil, errors.New("HubName Already exists")
+type chaosHubService struct {
+	ChaosHubOperation dbSchemaChaosHub.Operator
+}
+
+func NewService() Service {
+	chaosHubService := &chaosHubService{
+		ChaosHubOperation: dbSchemaChaosHub.NewOperator(),
 	}
 
-	cloneHub := model.CloningInput{
-		ProjectID:     chaosHub.ProjectID,
-		RepoBranch:    chaosHub.RepoBranch,
-		RepoURL:       chaosHub.RepoURL,
-		HubName:       chaosHub.HubName,
-		IsPrivate:     chaosHub.IsPrivate,
-		UserName:      chaosHub.UserName,
-		Password:      chaosHub.Password,
-		AuthType:      chaosHub.AuthType,
-		Token:         chaosHub.Token,
-		SSHPrivateKey: chaosHub.SSHPrivateKey,
+	// go routine for syncing hubs for all users
+	go chaosHubService.RecurringHubSync()
```

Comment:
Possible Bug, this function is called in 2 places :
1. GQL Resolver
2. ProjectInitializer function

According to this implementation, the only place the go routine should be started is when the resolver is set but because ProjectIntializer also calls this function every time a new project is created duplicate goroutines will start which will lead to excessive memory consumption.

The best thing to do is to remove this goroutine call from here and put it directly in the `main()` like it was before. Having a goroutine start inside some obscure function call is probably not the best idea especially when there should only be 1 such routine in a given instance of the server.